realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1045,23 +1045,33 @@ private Scanner getFullStringScanner(InputStream in) {
         if (!RealmObject.isManaged(parentObject) || !RealmObject.isValid(parentObject)) {
             throw new IllegalArgumentException("Only valid, managed objects can be a parent to an embedded object.");
         }
+
         RealmObjectProxy proxy = (RealmObjectProxy) parentObject;
         long parentPropertyColKey = schema.getSchemaForClass(parentObject.getClass()).getColumnKey(parentProperty);
         RealmFieldType parentPropertyType = schema.getSchemaForClass(parentObject.getClass()).getFieldType(parentProperty);
+        RealmSchema schema = getSchema();
+        RealmObjectSchema parentObjectSchema = schema.getSchemaForClass(parentObject.getClass());
+        String className = schema.getSchemaForClass(clazz).getClassName();
+        String linkedType = parentObjectSchema.getPropertyType(parentProperty);
         Row embeddedObject;
-        switch(parentPropertyType) {
-            case OBJECT: {
-                // FIXME: Check type of link
-                long objKey = proxy.realmGet$proxyState().getRow$realm().createEmbeddedObject(parentPropertyColKey);
-                embeddedObject = getTable(clazz).getUncheckedRow(objKey);
+
+        switch (parentPropertyType) {
+            case OBJECT:
+                if (linkedType.equals(className)) {
```

Comment:
Does it make sense to lift this check outside the `switch`? I'm not 100% sure how it would affect the `default` branch here.

Reply:
yeah, lifting it out would not work for the default branch, the default branch catches things like `INTEGER` and `BOOLEAN`... Not sure what would happen in that case?